### PR TITLE
chore: bump `nixpkgs` to 23.05

### PR DESCRIPTION
**Pull request by ggreif**
This is needed for tracking `ic-hs` which did the jump recently. `nixpkgs-22.11` got deprecated too.

Wasmtime generates different errors on trapping now.

_Note_: This `nixpkgs` revision is in synchronised with `ic-hs`'.

Remains to be debugged:
- [x] why is `shell` failing in CI, but not on the M1 Mac? — because `unit` tests used a `musl`-based toolset for OCaml builds, without obvious reason.
- [ ] why is the hash updater failing?
- [x] why is `json` version 4.1 still included in the closure? — due to https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/ocaml-modules/logs/default.nix, but that is optional
- [x] setting `dontDetectOcamlConflicts = true;` in the `shell` derivation papers over the problem

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -83,8 +83,8 @@ let commonBuildInputs = pkgs:
     pkgs.ocamlPackages.ppxlib
     pkgs.ocamlPackages.ppx_blob
     pkgs.ocamlPackages.ppx_inline_test
-    pkgs.ocamlPackages.ocaml-migrate-parsetree
-    pkgs.ocamlPackages.ppx_tools_versioned
+    # pkgs.ocamlPackages.ocaml-migrate-parsetree
```
Comment (ggreif) [inline default.nix:86]: this looks like a backporting opportunity

**Review by nomeata — APPROVED** [review]
Would be good to have the shell built though

**Comment by ggreif** [comment]
> Would be good to have the shell built though

Sure, I'll debug, but if it remains mysterious then I'll leave an issue. Also I wonder if @crusso can enter `nix-shell` on this branch.

File: default.nix [hunk 2]
```diff
@@ -426,9 +423,9 @@ rec {
       '';
     };
 
-    unit = ocamlTestDerivation {
+    unit = testDerivation {
```
Comment (ggreif) [inline default.nix:426]: it doesn't seem necessary to run unit tests in `musl` (static) link environment. @nomeata do you agree? It still passes and won't pull in duplicate OCaml library versions into the `moc-shell`.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -40,6 +40,8 @@ jobs:
         nix-build --max-jobs 1 -A drun
         nix-store --gc --print-roots | grep /motoko/
 
+    - run: nix-store -q --tree $(nix-instantiate -A shell)
```
Comment (ggreif) [inline .github/workflows/test.yml:43]: this can be removed now

**Comment by ggreif** [comment]
> Would be good to have the shell built though

Yep this is fixed now. `unit` tests now use the `.so`-toolset for OCaml. Do you think there is a catch?

**Review by crusso — APPROVED** [review]
Tested on my Linux machine.

**Pull request merged**